[denonmarantz] new Denon / Marantz 2.0 binding

jwveldhuis opened the pull request:
This is a  OpenHAB 2 port of the existing 1.x Denon binding (created by @idserda). As the binding supports both Denon and Marantz receivers decided to change the name to  'denonmarantz'.

Improvements:
- support for Discovery (MDNS)
- defined channels (including an 'advanced' Channel to send any custom command)
- improved Telnet stability (although that has been addressed in the old binding)

Note that this binding is still work in progress. Need others to test this binding as I only own one Marantz receiver (SR5008) myself. That is the main reason for this PR, as I know some cleanup, fixing and documentation is required.

This is my first binding, and it took more time and effort than I'd hoped for.. ;)
Happy to learn how to further improve.

jwveldhuis commented:
Not sure why Jenkins / Travis builds failed? Something I can do about that?

jwveldhuis commented:
Build issues fixed, found the jar file of the binding so other can test now.

jwveldhuis commented:
Testing by several users has been done over the past weeks.
I need to add a `README.md`.

jwveldhuis commented:
@martinvw ready for review! 
Just squashed all commits and added "Signed-off-by", but Pull Approve keeps telling that it is needed, tried with and without addeing the `(github: handle)`.

jwveldhuis commented:
Binding is made available at the Eclipse Marketplace: https://marketplace.eclipse.org/content/denonmarantz-binding

triller-telekom commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/discovery/DenonMarantzDiscoveryParticipant.java:
```diff
@@ -0,0 +1,163 @@
+                    telnetEnable = false;
+                    logger.debug("We can access the HTTP API, disabling the Telnet mode by default");
+                }
+            } catch (Exception e) {
```
See above: Please do not catch `Exception`

jwveldhuis replied:
See above, please suggest a solution for this.

triller-telekom commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/discovery/DenonMarantzDiscoveryParticipant.java:
```diff
@@ -0,0 +1,163 @@
+        Matcher matcher = DENON_MARANTZ_PATTERN.matcher(service.getQualifiedName());
+        if (matcher.matches()) {
+            logger.debug("This seems like a supported Denon/Marantz AVR!");
+            this.serial = matcher.group(1).toLowerCase();
```
What if you detect more than one? By storing the serial inside a class variable you will overwrite it in case there are 2 or more results.

jwveldhuis replied:
Ahh, didn't realize that! Thought the Participant would get instantiated for reach result, which obviously isn't the case. Good catch, this could have resulted in a hard to debug issue.

triller-telekom commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/telnet/DenonMarantzTelnetClient.java:
```diff
@@ -0,0 +1,168 @@
+ * @author Jeroen Idserda
+ * @author Jan-Willem Veldhuis
+ */
+public class DenonMarantzTelnetClient extends Thread {
```
Please make it a `Runnable` and schedule it with the scheduler (`scheduler.submit`..)

jwveldhuis commented:
@triller-telekom thank you for reviewing! I need to find time to work on your feedback.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,360 @@
+        <category>SoundVolume</category>
+    </channel-type>
+    
+    <channel-type id="mainVolumeDB" advanced="true">
```
The original binding supported both, also other AVR bindings support both Dimmer 0-100 style and actual dB value style. Some like OpenHAB to match the exact value the AVR is presenting on the (on-screen) display, this is the dB value. 
Deliberately made these channels 'advanced' to avoid confusion for regular users.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,360 @@
+        <category>SoundVolume</category>
+    </channel-type>
+    
+     <channel-type id="zone2VolumeDB" advanced="true">
```
See above.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,360 @@
+        <category>SoundVolume</category>
+    </channel-type>
+    
+    <channel-type id="zone3VolumeDB" advanced="true">
```
Same.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/about.html:
```diff
@@ -0,0 +1,39 @@
+<body lang="EN-US">
+<h2>About This Content</h2>
+ 
+<p>&lt;<em>May 4, 2015</em>&gt;</p> 
```
I guess it was the default contents generated by the skeleton command. Will update it to the PR date.

jwveldhuis replied:
Correct, updated the date.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/about.html:
```diff
@@ -0,0 +1,39 @@
+        for informational purposes only, and you should look to the Redistributor's license for 
+        terms and conditions of use.</p>
+        <p><em>
+        <strong>FasterXML 2.3.2</strong> <br/><br/>
```
No, removed this section.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/handler/DenonMarantzHandler.java:
```diff
@@ -0,0 +1,208 @@
+    private void addZoneChannels() {
+        // only add when this is a new Thing instance (and doesn't have the 'zones' property set yet).
+
+        if (this.getThing().getProperties().containsKey("zones")) {
```
I'm not too sure this is a right approach..
My intention was:
Some Denon/Marantz models have a 2nd zone, and some have an additional 3rd zone. I only want to add the specific channels when the AVR has these zones.
In the Thing property 'zones' (not to be confused with config parameter) I store the number of configured zones after adding the channels. So I use the property as a 'guard' to detect whether the channels were already added before.

With all knowledge gained in the meantime, I think it's better to detect the number of zones during Discovery and, in case of manual adding the Thing, to let the user override the default zones parameter (2) when needed.

With the Telnet connection there is no (convenient) way to detect the number of zones.

Will try to rewrite this part.

triller-telekom replied:
Having the property as a guard actually seems like a HACK to me.

So yes, discovering the channels within the discovery sounds plausible.

For the Telnet way: Why can't you discover the zones with it? What is the problem there?

jwveldhuis replied:
With Telnet we can only watch/query/set state, we cannot query properties. In theory could try to query the `Z2` and `Z3` state. When it replies within a short timeframe we could detect it has those zones.

triller-telekom replied:
Na this sounds like a HACK to me too, better leave out zone detection via telnet then.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/DenonMarantzConnectionProperties.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    public void setMainVolumeMax(BigDecimal mainVolumeMax) {
+        this.mainVolumeMax = mainVolumeMax;
```
Leftover from Denon 1.x binding, I've removed the `MAX_VOLUME` constant, and initialize mainVolumeMax with 98 instead.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/DenonMarantzConnector.java:
```diff
@@ -0,0 +1,785 @@
+    }
+
+    /**
+     * Set up a telnet connection to the receiver AND fetch initial state over HTTP.
```
You're right. Initially it was like that. Then in Denon 1.x it got changed while I was working on this 2.0 binding. Reason: not all (newer) receivers support HTTP, so this behavior was causing problems.
Will update the comment.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/DenonMarantzConnector.java:
```diff
@@ -0,0 +1,785 @@
+        if (!httpClient.isStopped()) {
+            try {
+                httpClient.stop();
+            } catch (Exception e) {
```
Totally agree with your argument. 
However, it was automatically generated by Eclipse (add throws/catch), and the Jetty HttpClient does throw 'Exception': https://www.eclipse.org/jetty/javadoc/current/org/eclipse/jetty/util/component/AbstractLifeCycle.html#stop--
When I try to catch `IOException` (for example), the compiler is complaining that `Exception` isn't caught.
Any suggestions to deal with this?

triller-telekom replied:
Oh, I see. In this case if this third party library throws only `Exception` I guess you do not have any choice other than catching it. So in this exceptional case it is OK to catch it but in general one should catch a more specific one.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/DenonMarantzConnector.java:
```diff
@@ -0,0 +1,785 @@
+            cmd += "OFF";
+        } else if (command instanceof RefreshType) {
+            if (!config.isTelnet()) {
+                // refreshing does not work using HTTP command, it will be polled soon.
```
By default polling is done every 5 seconds, also with HTTP the refresh is done for ALL items as it fetches and processes an xml with the receiver state.
I've observed that upon start of OpenHAB all items are being refreshed. That would result in a lot of requests being executed "in between" the regular interval (probably these will all run in parallel and might even cause issues at the end of the receiver).

You are right about the need to factor this out, will do that.

triller-telekom replied:
> I've observed that upon start of OpenHAB all items are being refreshed

That is true. Upon the OH2 start a `Refresh` command is send to all items. For http this should be fine. But if you are saying that this would be a problem for the telnet connection because you will have a new poll attempt for each item in a very short time frame, then I guess we have to live with the fact that we have to wait until the normal polling fetches the data.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/DenonMarantzConnector.java:
```diff
@@ -0,0 +1,785 @@
+     */
+    public void refreshState() {
+        if (config.isTelnet()) {
+            requestStateOverTelnet();
```
Telnet doesn't support this..

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/DenonMarantzConnector.java:
```diff
@@ -0,0 +1,785 @@
+     *
+     * @param line The received command (one line)
+     */
+    // TODO implement/update telnet command -> state update
```
:) That was done in the meantime, forgot to remove the comment.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/DenonMarantzConnector.java:
```diff
@@ -0,0 +1,785 @@
+     * @throws IOException
+     */
+    public void refreshState() {
+        if (config.isTelnet()) {
```
I agree, will first fix the other findings in order to not break things.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/DenonMarantzConnector.java:
```diff
@@ -0,0 +1,785 @@
+     */
+    private void requestStateOverTelnet() {
+        // Run in new thread to not delay the set up of the binding. Handling responses is done in the listener thread.
+        new Thread(new Runnable() {
```
Changed.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/DenonMarantzConnector.java:
```diff
@@ -0,0 +1,785 @@
+
+        } catch (UnsupportedEncodingException e) {
+            logger.warn("Error sending command", e);
+        } catch (Exception e) {
```
See above (`Exception` is thrown by HttpClient.start())

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/DenonMarantzState.java:
```diff
@@ -0,0 +1,259 @@
+        if (!newVal.equals(this.mainVolume)) {
+            this.mainVolume = newVal;
+            listener.stateChanged(DenonMarantzBindingConstants.CHANNEL_MAIN_VOLUME, this.mainVolume);
+            // update the main volume in dB too
```
See my previous explanation.
See https://github.com/openhab/openhab2-addons/tree/master/addons/binding/org.openhab.binding.yamahareceiver
and https://github.com/openhab/openhab2-addons/tree/master/addons/binding/org.openhab.binding.pioneeravr

triller-telekom replied:
If people wish to have the "dB" channel alongside the volume channel and it has been implemented in the other avr bindings too, I am fine with keeping it.

jwveldhuis commented:
Note to self: need to fix issue in 2.3.0 caused by this PR: 2997
(I used the Yamaha binding as a reference for adding Channels dynamically)

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/handler/DenonMarantzHandler.java:
```diff
@@ -0,0 +1,208 @@
+    @Override
+    public void stateChanged(String channelID, State state) {
+        logger.debug("Recieved state {} for channelID {}", state, channelID);
+        if (this.getThing().getStatus() != ThingStatus.ONLINE) {
```
Removed this.
Did apply similar check for setting the Thing to OFFLINE only once though. As it can be set offline with different descriptions it was causing a flood in the event log in some circumstances (see https://community.openhab.org/t/denon-binding-warn-log-messages/33024/12)

jwveldhuis replied:
Re-added it after I saw the events.log getting filled with `'denonmarantz:avr:15f35735' updated: ONLINE`. If any Channel state changes, it goes through this function.
We only want to notify the system about the Thing's online status when it actually changed to ONLINE.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/DenonMarantzConnector.java:
```diff
@@ -0,0 +1,785 @@
+            throw new UnsupportedCommandTypeException("Zone must be in range [1-3], zone: " + zone);
+        }
+        String zonePrefix = zone == 1 ? "" : "Z" + zone;
+        String cmd = zonePrefix + "MU";
```
Done.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,360 @@
+
+            <parameter name="httpPollingInterval" type="integer" required="true" groupName="httpSettings">
+                <label>Polling interval</label>
+                <description>Refresh interval of the HTTP API in milliseconds</description>
```
Done. Should I handle an automatic conversion from 5000 to 5 for all current (BETA) users of the binding?

triller-telekom replied:
No I think it doesn't make much sense to clutter the code with unnecessary conversion logic. BETA users should be aware oft hat it was a BETA and thus subject to be changed. maybe you can leave them a note in the market place or where ever you posted your BETA version.

jwveldhuis commented:
@triller-telekom how do I handle the merge with upstream/master in between, should that be 'signed-off' as well?
Sorry I'm not too familiar with git / github.

triller-telekom commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/DenonMarantzState.java:
```diff
@@ -58,6 +58,10 @@ public DenonMarantzState(DenonMarantzStateChangedListener listener) {
     }
 
+    public void connectionError(String errorMessage) {
+        listener.connectionError(errorMessage);
```
What if the `listener` is `null`?

jwveldhuis replied:
It can't be. State is instantiated from the `DenonMarantzHandler` like this, the handler is the (only) listener.
```java
denonMarantzState = new DenonMarantzState(this);
```

triller-telekom replied:
Yes I saw that is is initialized with the handler, but code gets changed over time and who knows if there won't be another instance created somewhere in the future? Just saying this because with the term `listener` I usually connect that it is a 0-n relationship, i.e. it is optional or there may be many.

jwveldhuis replied:
I agree, will rename the variable to `handler` to avoid future confusion.

triller-telekom commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/connector/http/DenonMarantzHttpConnector.java:
```diff
@@ -0,0 +1,399 @@
+
+    private void startPolling() {
+        if (!isPolling()) {
+            logger.info("HTTP polling started.");
```
Please decrease to debug level.

triller-telekom commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/connector/http/DenonMarantzHttpConnector.java:
```diff
@@ -0,0 +1,399 @@
+    private void stopPolling() {
+        if (isPolling()) {
+            pollingJob.cancel(true);
+            logger.info("HTTP polling stopped.");
```
Please decrease to debug level.

triller-telekom commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/connector/http/DenonMarantzHttpConnector.java:
```diff
@@ -0,0 +1,399 @@
+    }
+
+    private void updateSecondaryZones() {
+        for (int i = 2; i <= config.getZoneCount(); i++) {
```
Why are you using a loop at all if you have a switch-case later on anyway?

jwveldhuis replied:
To reuse the code to construct the url, to get the document and to log about that.

triller-telekom commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/connector/http/DenonMarantzHttpConnector.java:
```diff
@@ -0,0 +1,399 @@
+        String ret = IOUtils.toString(is);
+
+        connection.disconnect();
+
```
empty line

triller-telekom commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/discovery/DenonMarantzDiscoveryParticipant.java:
```diff
@@ -99,27 +128,88 @@ public DiscoveryResult createResult(ServiceInfo serviceInfo) {
                         .timeout(3, TimeUnit.SECONDS).send();
-                int status = response.getStatus();
-                if (status == 200) {
+                if (response.getStatus() == 200) {
```
Isn't there a constant like `HTTP_STATUS.OK` or something instead of the 200?

triller-telekom commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/discovery/DenonMarantzDiscoveryParticipant.java:
```diff
@@ -99,27 +128,88 @@ public DiscoveryResult createResult(ServiceInfo serviceInfo) {
+                try {
+                    response = httpClient.newRequest("http://" + host + ":8080/goform/Deviceinfo.xml")
+                            .timeout(3, TimeUnit.SECONDS).send();
+                    if (response.getStatus() == 200) {
```
See above

triller-telekom commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/discovery/DenonMarantzDiscoveryParticipant.java:
```diff
@@ -99,27 +128,88 @@ public DiscoveryResult createResult(ServiceInfo serviceInfo) {
+                    logger.debug("Failed in fetching the Deviceinfo.xml to determine zone count: {}", e.getMessage());
+                }
+
+                if (status == 200 && response != null) {
```
See above

jwveldhuis commented:
@triller-telekom ahh, thanks for this basic git-HOWTO :) Searched and tried a lot, couldn't find this.
Will work on your latest comments.
And indeed, the code it a lot cleaner now. When I started this I wanted to keep as close the the 1.x binding, now eventually almost everything got changed.. ;)

jwveldhuis commented:
@triller-telekom what is the recommended pattern to handle `RunTimeException` in a Runnable scheduled with `ScheduledExecutorService`?
Learned today that the `Runnable` just stops and NO error is logged anywhere.. :/
See https://stackoverflow.com/a/24902026/2486310 

I am planning to handle it like this:
```java
pollingJob = scheduler.scheduleWithFixedDelay(new Runnable() {
    @Override
    public void run() {
        try {
            refreshHttpProperties();
        } catch (Throwable t) {
            StringBuilder sb = new StringBuilder();
            for (StackTraceElement s : t.getStackTrace()) {
                sb.append(s.toString()).append("\n");
            }
            logger.error("Error while polling Http: \"{}\". Stacktrace: \n{}", t.getMessage(),
                    sb.toString());
        }
    }
}, config.httpPollingInterval, config.httpPollingInterval, TimeUnit.SECONDS);
```

triller-telekom commented on addons/binding/org.openhab.binding.denonmarantz/META-INF/MANIFEST.MF:
```diff
@@ -1,32 +1,36 @@
- org.eclipse.smarthome.io.transport.mdns.discovery,
  org.openhab.binding.denonmarantz,
  org.openhab.binding.denonmarantz.handler,
+ org.osgi.service.component.annotations;resolution:=optional,
```
Are you sure that you removed `org.osgi.service.component.annotations;resolution:=optional,` and **NOT** accidentally `org.eclipse.jdt.annotation;resolution:=optional,`? Because the later one would explain your error message...

jwveldhuis replied:
Mea culpa, yes I did.
However, when I remove the `org.osgi.service.component.annotations;resolution:=optional,` I see the following warning in Eclipse:
`DS Annotations missing from permanent build path`

triller-telekom replied:
This is because your branch is too old. I have just tested a plain checkout with the current master and everything is fine in the IDE. if I rebase your branch onto the current master and remove `org.osgi.service.component.annotations;resolution:=optional,` from the `MANIFEST.MF`, everything is alright.

So please rebase your branch onto the current master and do a push with --force. Careful though, because this will overwrite the branch on github. So you might want to do a backup of your current local repository before doing so since you said you had problems with git before.

However, I just did a `git rebase master` in your branch locally (after pulling the latest master branch of course) on my machine and everything went smooth.

jwveldhuis replied:
I am on the latest commit (commit dc204c6cc8fc3e1be9296c773b0c4eade3c4f7aa) of openhab2-addons, still see the warning.
Is it perhaps a setting outside of this repo? What was changed to resolve this warning?

jwveldhuis replied:
Nevertheless, I will apply the change and ignore the local warning (I see them for other bindings as well), perhaps related to my IDE setup.

triller-telekom replied:
All of you checked out repos have to be on the newest version, i.e. "openhab1-addons  openhab2-addons  openhab-core  openhab-distro  smarthome"

jwveldhuis replied:
I only have `openhab2-addons` and `openhab-distro` and both are up to date (while [setting up the IDE](https://docs.openhab.org/developers/development/ide.html#installation) I only picked "openHAB Development" (required) and "openHAB2 Add-ons" from the optional set of repos).

jwveldhuis replied:
Confirmed warning is gone when all git repos are included in the IDE.

triller-telekom commented:
Regarding your `RuntimeException` in a `Runnable` problem: We had https://github.com/eclipse/smarthome/pull/4772

But it turned out that [some jobs were not executed](https://github.com/eclipse/smarthome/issues/4931) and thus we had to revert it via https://github.com/eclipse/smarthome/pull/4932

I just tried to play around with it but unfortunately I could not really reproduce it :( Neither in my IDE pure ESH setup nor in a basic installation of OH 2.2 with a patched smarthome.core bundle (one that contained the wrappedscheduler).

So for now I guess you have to stick to catching the `RuntimeException` on your own. If you found out some more details while you where near the code in your debug session anyway, help from your side is more than welcome :)

triller-telekom commented:
@jwveldhuis FYI: With https://github.com/eclipse/smarthome/pull/5045 I have fixed the logging of `Exception`s in scheduled jobs :)

If you address the few minor issues from above we can handover your PR to the committers for a final review and merge.

jwveldhuis commented:
@triller-telekom excellent! Thank you!
Need to find time to finish the last issues.
Also started to work on dynamically fetching the Input options (e.g. TV, NET, TUNER). I will postpone that for a follow-up PR to not further delay this.

jwveldhuis commented:
@triller-telekom does your fix only address the error logging, or does it also keeps the scheduler running for subsequent tasks (I don't think so)?
So I think I still need to catch `RuntimeException` to prevent the scheduler from stopping.

triller-telekom commented:
> So I think I still need to catch RuntimeException to prevent the scheduler from stopping. Will perform some testing around this.

Yes indeed. My fix is only to make the error show up in the logs. if you already know that a `RuntimeException` may occur and how to deal with it, then you can catch it. So my fix doesn't change any behavior it just makes a wrong behavior obvious.

jwveldhuis commented:
Pushed changes.
Again struggled with git. I rebased to master, but then couldn't push:
```
 ! [rejected]          denon-marantz-binding -> denon-marantz-binding (non-fast-forward)
error: failed to push some refs to 'https://github.com/jwveldhuis/openhab2-addons.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
```
In `git log` I saw my previous commits got a different commit hash due to the `git rebase`. Would GitHub still recognise the previous commits or would it mess up the code review? I now reverted back to the previous state of my fork and then put the changes on top, then pushed.

jwveldhuis commented:
@triller-telekom done.

triller-telekom commented:
@kaikreuzer @martinvw This PR is now ready for a final review and can be merged afterwards. Please have a final look at it. Thanks.

triller-telekom commented:
Thank you very much @jwveldhuis !

jwveldhuis commented:
Thank you for the extensive review! Enjoyed the collaboration!
Once this is merged will come back with some more enhancements ;)

kaikreuzer commented:
Thanks for this contribution @jwveldhuis!
I just started with doing a final review. What immediately caught my eye is that you do not seem to be familiar with the concept of [channel groups](https://www.eclipse.org/smarthome/documentation/development/bindings/thing-definition.html#channel-groups) - those are actually an ideal fit for the different zones of an AVR, which all have the very same features (= set of channels). You'd therefore only have to define all those channels once and then add three groups for the three zones. Would that make sense?

jwveldhuis commented:
@kaikreuzer Yes, makes sense.. Will look into that.
Question about this: Later on, I want to dynamically update the input channel(s) (to change the list of StateOptions with the available inputs). Would that become an issue when using channel groups? Do I then need to update the whole set of channels?

kaikreuzer commented:
No, not at all - all you have to do is to implement a [DynamicStateDescriptionProvider](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/type/DynamicStateDescriptionProvider.java), which can tweak any option list dynamically at runtime.

jwveldhuis commented:
Had a look at channel-groups, but I'm not convinced that it will save my or future developers time. It will take hours to rewrite and test. I noted this in the doc:

> A thing can only have direct channels or channel groups, but not both.

Meaning I should create separate channel groups for the AVR (power), the mainZone (to distinguish main power from main zone power) and the optional 2nd and 3rd zone. Only for the 2nd and 3rd zone a little efficiency could be gained because they share the same channel types. Zone channels are in the minority compared to general AVR and mainZone channels. In order to provide the same quality of the labels / descriptions (like `Input source (Zone 2)`, `Select the input source for Zone 3 of the AVR`) some mechanism would be needed to override the labels and descriptions for these channels (and then hence losing all the efficiency).

kaikreuzer commented:
> Zone channels are in the minority compared to general AVR and mainZone channels.

Not really. I only see power/surroundProgram/artist/album/track/command as AVR channels, while the other 15 (3x5) channels are zone channels.

> some mechanism would be needed to override the labels and descriptions for these channels 

Note that the labels/descriptions should not depend on the zone, but they should describe the functionality (which is identical for all 3 zones). In the Thing view it will be clear to which zone the channel belongs two, because they will be shown grouped by zone.

jwveldhuis commented:
Ok, maybe it is. Already invested a lot of time in this binding. So if someone else wants to (help) refactor it to use `channel-groups` please go ahead. It's not an easy change for me, don't know where to start to change everything (adding/removing the zone channels, etc).

jwveldhuis commented:
@kaikreuzer any update on the review? Apart from recommendation to use channel groups (which is too hard / time consuming for me atm to apply), is there anything else that needs to be fixed before it can be released?

kaikreuzer commented:
@jwveldhuis I actually don't think that the refactoring to channel-groups would mean a huge work - it mainly removes a lot of redundancy and will require a renaming of the channel id constants.
As you own the hardware and have written the code, I'd very much hope that you'd be willing to do this task. We can alternatively also add a "help wanted" label to the PR, but I doubt that we will quickly find someone taking this over. 
Note that merging first and refactoring later isn't really an option as it would mean a breaking change for users, which we should definitely avoid. Furthermore, all code in the official repo serves as an example for others - that's why it often causes additional confusion and efforts, if such temporary solutions are merged.

I very much appreciate all the efforts you have put into it already and would hope that you are willing to continue on the last mile as well. If you agree, I'd schedule this binding for a final review (there shouldn't be much coming up from it, no worries) asap. Let me know what you think!

jwveldhuis commented:
@kaikreuzer ok I agree, refactoring after releasing does imply an API breaking change.
Let me see when I can start running the last mile. After all I do want this to be completed after so much work!

kaikreuzer commented:
Great, thanks for you understanding - let's get that done 👍 .

jwveldhuis commented:
@kaikreuzer how do I dynamically add channel groups to the Thing? 
This is in the `thing-types.xml` now:
```
<channel-groups>
  <channel-group id="general" typeId="general"/>
  <channel-group id="mainZone" typeId="zone"/>
</channel-groups>
```
How do I dynamically add the channel-group `zone2` of type `zone` including the corresponding channels? Any existing binding doing this?

jwveldhuis commented:
Or should I assign all available channel groups and remove the specific channels which are not applicable?
E.g. assign zone2 and zone3 channels and remove those during initialization when it turns out the receiver doesn't support those?

kaikreuzer commented:
@jwveldhuis Why do you need to add the zones dynamically? Isn't it clear, which zones are available for a specific receiver model? Ideally, you have different thing types for them, which statically define the available zones.

If that isn't possible (possibly because the set of main channels also differ between models), you would indeed have to dynamically add them. There isn't yet any specific support in the framework for adding all channels of a group at once to a Thing, so you would have to add the channels one by one. Note that the Thing itself does not hold any information about channel groups (they only exist on the Thing type). When adding channels on the Thing, you would simply add channels with the id "zoneX#volume" with channel type id "volume" etc.
How channels are dynamically added in general is described [here](https://www.eclipse.org/smarthome/documentation/development/bindings/thing-handler.html#updating-the-thing-structure).

kaikreuzer commented:
> by dynamically detecting the number of zones we avoid having to update the binding

ok, fine.

> I guess I do need to assign all channel-groups upfront

As mentioned above: There is no information about channel groups at all in the Thing, so there is no need to create or assign any. All that exists as a reference is the channel id with 'zoneX#volume' format (in constrast to the normal 'volume' channel id).

jwveldhuis commented:
@kaikreuzer I start to get it now.. Struggling with the code which dynamically adds / removes channels. Was using the ChannelTypeUID as a unique identifier to add/remove channels, this obviously does no longer work. Need to think of an alternative mechanism. 

Big drawback of using channel groups is the lack of custom labels / descriptions for each channel.
How to override the labels / descriptions? (I guess that defeats the purpose of channel groups?)
In Paper UI it is becoming messy for users now the custom channels for each zone are replaced by generic ones:

- Zone control
  - Power (zone)
  - Volume
  - Mute
  - Input source
- Zone control
  - Power (zone)
  - Volume
  - Mute
  - Input source
- Zone control
  - Power (zone)
  - Volume
  - Mute
  - Input source

How to distinguish Main Zone from Zone1 and Zone2? I used to have custom labels, e.g. "Power (zone2)", "Mute (zone3)", etc.

jwveldhuis commented:
Just pushed the updated Binding. Have created separate channel groups in order to provide a custom description for each channel group (while still reusing the common channels).

kaikreuzer commented:
Sorry for my late reply - there actually is no need to define multiple channel-group-types. For custom labels&descriptions, you can simply overwrite them for each channel group:
```
<channel-groups>
    <channel-group typeId="zone" id="mainZone">
        <label>Main Zone</label>
        <description>The main zone of this AVR</description>
    </channel-group>
    <channel-group typeId="zone" id="zone2">
	    <label>Zone 2</label>
        <description>Zone 2 of this AVR</description>
    </channel-group>
    ...
</channel-groups>
```
This should simplify you xmls again a bit (and hopefully not require any code changes at all).

jwveldhuis commented:
@kaikreuzer ah thanks, now it all makes sense 😁 
Just pushed the suggested change.

kaikreuzer commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/connector/http/DenonMarantzHttpConnector.java:
```diff
@@ -0,0 +1,410 @@
+                public void run() {
+                    try {
+                        refreshHttpProperties();
+                    } catch (RuntimeException e) {
```
I just saw your discussion and your comment

> So I think I still need to catch RuntimeException to prevent the scheduler from stopping. Will perform some testing around this.

I hope your testing showed that it is ok to stop the scheduled job here - if the problem is due to a configuration error, the job should be rather re-scheduled once the configuration is fixed.

jwveldhuis replied:
Well the main reason was the exception was nowhere to be found and was very very difficult to diagnose and debug due to this. What if 20 channels are fine and only 1 channel is mistakenly misconfigured (e.g. Dimmer type used for Item instead of Number type), not sure whether we should stop everything? At least the error should be catched and logged..

How to determine whether the configuration has been fixed so polling can resume?

jwveldhuis replied:
Could I use the `channelLinked` as trigger to resume polling? E.g. when one has updated an Item definition? Reconfiguration of the Thing itself currently restarts everything already.
The only known cause until now for the RunTimeException was a malformed manual Item definition in a `.items` file.

jwveldhuis replied:
BTW is there a way to warn the user he/she has mapped a `Dimmer` item to a `Number` channel? It can work for a while, but at some point you can get nasty errors. The error thrown is not very helpful in telling the user it was their fault and they should look into the item definition.

jwveldhuis replied:
@kaikreuzer This remains an open item. Any suggestion on how to trigger a resume of the polling after it failed?

kaikreuzer replied:
Sorry, @jwveldhuis, I missed to follow up on your questions.

> BTW is there a way to warn the user he/she has mapped a Dimmer item to a Number channel? 

No, I actually consider it a bug in the framework that it is at all possible (and that handlers are affected by it), see https://github.com/eclipse/smarthome/issues/2666.

To come to a solution here: Let's blame the framework for throwing RuntimeExceptions here and let's keep the code as it is for now.

idserda commented:
Great work! Not sure how much of my original code is still left but you definitely have my approval.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/handler/DenonMarantzHandler.java:
```diff
@@ -0,0 +1,275 @@
+        }
+
+        if (connector instanceof DenonMarantzHttpConnector && command instanceof RefreshType) {
+            // Refreshing individual channels isn't supported by the Http connector.
```
I know, but the AVR can't handle too many HTTP requests at once. I had to power cycle my AVR during tests when all linked channels got polled during startup.

kaikreuzer replied:
Sure. Just fyi: We have a [expiring cache](https://www.eclipse.org/smarthome/documentation/features/frameworkUtilities.html#caching) that is meant for this situation (reducing the number of calls that go out to the device).

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/connector/http/DenonMarantzHttpConnector.java:
```diff
@@ -0,0 +1,410 @@
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import javax.xml.bind.JAXBContext;
```
Was already used in the 1.x binding, will investigate whether it can be easily swapped with xstream.

jwveldhuis replied:
After checking out xstream documentation and blogs about converting JAXB to xstream I feel it is too much work to rewrite and especially retest everything (I only own 1 AVR which does not produce all supported XML). The XML processing is mainly untouched with this 2.x upgrade. Some advanced features are used (like XmlAdapter) for which a whole new Converter implementation is required in xstream.
So I prefer to stay with JAXB and will add the javax packages to the manifest.

jwveldhuis commented on addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/discovery/DenonMarantzDiscoveryParticipant.java:
```diff
@@ -0,0 +1,237 @@
+
+            logger.debug("IP Address: {}", host);
+
+            // try a HTTP request to autoconfigure the HTTP / Telnet parameter
```
I agree it's not perfect. For some models HTTP is not an option, so for those I want to set 'Use Telnet' to `true` in the Thing parameter. The default for this parameter is `false`.

Should I do this instead?
- set some property in the Thing indicating it was created via a Discovery Result (I guess there is no way to determine otherwise?)
- if set, auto-detect and alter the value for parameter 'Use Telnet' during Inititialization

E.g. I don't want to override the parameter when "explicitly" (there is a default) set by the user during manual Thing creation (for whatever reason).

jwveldhuis replied:
I found another option, but have trouble implementing this:
Limit the number of times the procedure is invoked to exactly once, by stopping further (relatively expensive) discovery activities as soon as it's clear we already configured or discovered the Thing.
https://www.eclipse.org/smarthome/documentation/development/bindings/discovery-services.html#compare-with-existing-results

I tried to implement the `ExtendedDiscoveryService` with the `MDNSDiscoveryParticipant`, this does not work unfortunately. How can I add this callback to the MDNSDiscoveryService? How do I approach this?

triller-telekom replied:
@jwveldhuis This is a missing feature which I have just implemented in #5526

jwveldhuis replied:
@triller-telekom perfect!

kaikreuzer replied:
I am not sure whether this would be a solution. After all, you still have to do HTTP requests in a method that must return fast. And discovery is supposed to be passive in most cases, it should really just sit there and listen.

Couldn't you simply leave this property unset (=null) and upon handler initialisation, you do the HTTP call and figure out the right value, if it is still null? This would allow you to set the value without bothering the user and still avoid the risk of overwriting a value that the user has specifically set.

jwveldhuis replied:
Ok, I have now moved this part over the the Handler. Added benefit is the auto configuration for telnet/http and zone count is now also applied to Things configured in text files.

grzegorz914 commented:
Hi,

I use Your binding wit OH2.4 and AVR-X6300H, all working OK except "surroundProgram", this tag show nothing, connection HTTP or Telnet make no differences. Look on screenshot.

<img width="1576" alt="zrzut ekranu 2018-07-11 o 18 15 23" src="https://user-images.githubusercontent.com/28867813/42585832-f18e4c6a-8536-11e8-8add-37433fbdfeca.png">

Thanks
Grzegorz

jwveldhuis commented:
@grzegorz914 please create a new issue for this. Please don’t add comments to closed issues.
